src/object/cli_obj.c

```diff
@@ -2622,3 +2661,457 @@ dc_obj_query_key(tse_task_t *api_task)
+		}
+
+		/* We used stale pool map, need to check new leader. */
+		if (rc == -DER_STALE)
+			goto leader;
+
+		/* Enlarge the buffer and re-list the dkeys. */
+		if (rc > 0)
+			goto list;
+
+		if (rc < 0) {
+			if (rc != -DER_NONEXIST)
+				D_GOTO(out, rc);
+
+			rc = 0;
+			/* Leader replica has been destoryed, need to check
+			 * whether other replicas are there or not.
+			 */
+			arg_l->lost_shard = 1;
+		} else {
+			arg_l->lost_shard = 0;
+		}
+
+		if (arg_l->has_retried && !arg_l->lost_shard &&
+		    arg_l->size == arg_l->size_saved &&
+		    memcmp(arg_l->buf, arg_l->buf_saved, arg_l->size) == 0) {
+			/* Nothing changed before and after retry with the
+			 * leader, then the leader must be insistent with
+			 * at least one the of the other replicas.
+			 */
+			D_INFO(DF_OID" (reps %d, inconsistent) between shard "
+			       "%d (l) and shard %d, may be different keys\n",
+			       DP_OID(oid), reps, leader + start, i + start);
+			D_GOTO(out, rc = -DER_MISMATCH);
+		}
+
+		for (i = 0; i < reps && i != leader; i++) {
+			struct obj_enum_dkeys_arg	*arg_s = &oeda[i];
+
+			if (arg_s->buf_len < arg_l->buf_len) {
+				if (arg_s->buf != arg_s->inline_buf)
+					D_FREE(arg_s->buf);
+
+				arg_s->buf_len = arg_l->buf_len;
+				D_ALLOC(arg_s->buf, arg_s->buf_len);
+				if (arg_s->buf == NULL)
+					D_GOTO(out, rc = -DER_NOMEM);
+			}
+
+			rc = daos_enum_dkeys_do_list(oh, &epoch, arg_s, list_cb,
+						     DIOF_TO_SPEC_SHARD);
+
+			if (arg_s->need_retry) {
+				arg_s->need_retry = 0;
+				arg_s->has_retried = 1;
+			}
+
+			/* We used stale pool map, need to check new leader. */
+			if (rc == -DER_STALE)
+				goto leader;
+
+			/* If non-leader has more things than leader, then:
+			 * 1. The leader and non-leader may be inconsistent. Or
+			 * 2. There are some modifications committed after list
+			 *    from the (new) leader.
+			 *
+			 * We need to re-list the dkeys with leader. If leader
+			 * has more things, it may be the 2nd case. Otherwise,
+			 * it is the 1st case.
+			 */
+			if (rc > 1) {
+				int	j;
+
+				oeda[leader].need_retry = 1;
+				for (j = 0; j < i; j++)
+					oeda[j].need_retry = 1;
+
+				goto list;
+			}
+
+			if (rc < 0) {
+				if (rc != -DER_NONEXIST)
+					D_GOTO(out, rc);
+
+				rc = 0;
+				arg_s->lost_shard = 1;
+			} else {
+				arg_s->lost_shard = 0;
+			}
+
+			if (arg_l->lost_shard) {
+				if (arg_s->lost_shard)
+					continue;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) does not exists, but shard %d "
+				       "does exists.\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (arg_s->lost_shard) {
+				/* Non-leader does not exist, but leader seems
+				 * there. It may because someone destroyed the
+				 * object after we list from leader. Let's try
+				 * to re-list from leader.
+				 */
+				if (!arg_l->has_retried) {
+					int	j;
+
+					oeda[leader].need_retry = 1;
+					for (j = 0; j < i; j++)
```

Comment:
why not is  "j <= i" in the for?

Reply:
The shard[i] is the current replica that has different (may be newer) data than former replicas, we only need to re-list/re-fetch from former replicas (including the leader), but not need to re-list/re-fetch from current replica.

Reply:
ok, thx